profiling/src/lib.rs

```diff
@@ -336,6 +307,92 @@ pub struct RequestLocals {
     pub vm_interrupt_addr: *const AtomicBool,
 }
 
+/// take a sample every X bytes
+const ALLOCATION_PROFILING_INTERVAL: f32 = 1024.0 * 30.0;
+
+pub struct AllocationProfilingStats {
+    /// number of bytes in "this" sampling interval
+    pub sampling_interval: AtomicU64,
+    /// number of bytes until next sample collection
+    pub remaing_bytes: AtomicI64,
+    /// number of allocations since last sample
+    pub allocation_count: AtomicU64,
```

Comment:
A lot of the implementation seems to deal with atomics. FWIW Go maintains one `nextSample` (similar to what you call remaining_bytes) for each thread. This means there is no need to deal with atomics. Each threads ends up making its own independent sampling decisions. Maybe this makes sense for your PR as well?